lib/handler/RedirectHandler.js

```diff
@@ -24,13 +24,18 @@ class BodyAsyncIterable {
 }
 
 class RedirectHandler {
-  static buildDispatch (dispatcher, maxRedirections) {
-    if (maxRedirections != null && (!Number.isInteger(maxRedirections) || maxRedirections < 0)) {
-      throw new InvalidArgumentError('maxRedirections must be a positive number')
-    }
+  static decorate (dispatch) {
```

Comment:
Early review, feel free to ignore it if you already considered it.

Should we consider the `decorate` to pass custom options?
I see potential there as it can configure the _decorator_ and pass it down to the actual decoration, and maybe even allow to be overwritten at the moment of using it.

Maybe won't be needed for the `RedirectHandler`, but it will for the `Retry` one, wdyt?

Reply:
This is just an experiment PR. Feel free to shoot all ideas. I'll have an update later this week.

Reply:
Sure, this is what I was having in mind:

```js
static decorate (dispatch, handlerOpts) {
  // We validate the options for the handler are valid
  // These will be the defaults if not passed through the dispatch call
  if (handlerOpts?.maxRedirections == null || handlerOpts.maxRedirections === 0) {
    return dispatch(opts, handler)
  }

  // By splitting it into two phases, we allow them to override the globals if they need to
  // This can also showcase how user-land can create its own custom implementations
  // and even pass custom options to it (e.g. RetryHandler)
  return (opts, handler) => {
    const redirectHandler = new RedirectHandler(dispatch, handlerOpts);
    return dispatch(opts, redirectHandler.handle(opts, handler))
  }
}
```

Reply:
how that would be used?

Reply:
There's when I go over the fence; sadly this will require implementers to do use the `dispatch` to make it work.
Or, something like you did, create a custom Agent on top (`RetryAgent`).

Maybe this is something good and we should strive implementers to seek for to implement custom behaviours in that way?

Reply:
I kind of like @mcollina's pattern in RetryAgent etc. Will give that more thought. My main concern with the current interceptor design is that unnecessarily lives inside the core instead of just wrapping it.